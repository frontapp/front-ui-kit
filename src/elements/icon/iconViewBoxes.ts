// Generated metadata for original SVG viewBox values
// Only stores exceptions - most icons use the default "0 0 16 16"
// This ensures we can preserve the original viewBox from SVG files
// when no viewBox prop is explicitly provided

import type { IconName } from './icon';

const ICON_VIEW_BOX_EXCEPTIONS: Partial<Record<IconName, string>> = {
  Export: '0 0 16 17',
  Info20: '0 0 20 20',
  CaretExpand: '0 0 20 20',
  EditSquare: '0 0 20 20',
  GearRounded: '0 0 18 20'
};

const DEFAULT_VIEW_BOX = '0 0 16 16';

/**
 * Gets the original viewBox for an icon name.
 * Returns the exception viewBox if one exists, otherwise returns the default.
 */
export function getIconViewBox(iconName: IconName): string {
  return ICON_VIEW_BOX_EXCEPTIONS[iconName] || DEFAULT_VIEW_BOX;
}
