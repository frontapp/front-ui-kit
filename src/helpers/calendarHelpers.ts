import {range} from 'lodash';
import {DateTime} from 'luxon';

/*
 * Constants.
 */

export enum CalendarWeekDaysEnum {
  SUNDAY = 'sunday',
  MONDAY = 'monday',
  TUESDAY = 'tuesday',
  WEDNESDAY = 'wednesday',
  THURSDAY = 'thursday',
  FRIDAY = 'friday',
  SATURDAY = 'saturday'
}

export const calendarWeekdayToWeekdayNumber = {
  [CalendarWeekDaysEnum.MONDAY]: 1,
  [CalendarWeekDaysEnum.TUESDAY]: 2,
  [CalendarWeekDaysEnum.WEDNESDAY]: 3,
  [CalendarWeekDaysEnum.THURSDAY]: 4,
  [CalendarWeekDaysEnum.FRIDAY]: 5,
  [CalendarWeekDaysEnum.SATURDAY]: 6,
  [CalendarWeekDaysEnum.SUNDAY]: 7
};

/*
 * Helpers.
 */

/** Get the sorted weekdays number, starting with Sunday. */
export function getSortedWeekdays(calendarWeekStartDay: CalendarWeekDaysEnum) {
  const sortedCalendarWeekdays = getSortedWeekdayNames(calendarWeekStartDay);
  return sortedCalendarWeekdays.map(weekdayName => calendarWeekdayToWeekdayNumber[weekdayName]);
}

export function getSortedWeekdayNames(calendarWeekStartDay: CalendarWeekDaysEnum) {
  const unorderedWeekdays = Object.values(CalendarWeekDaysEnum);
  const splitIndex = unorderedWeekdays.indexOf(calendarWeekStartDay);
  const left = unorderedWeekdays.slice(0, splitIndex);
  const right = unorderedWeekdays.slice(splitIndex);
  return [...right, ...left];
}

/** Get the full list of days for the provided month, sorted by the provided week days. */
export function getCalendarDays(month: DateTime, weekdays: ReadonlyArray<number>) {
  // Find the first datetime of the months we're going to work with.
  const previousMonth = month.minus({months: 1});
  const nextMonth = month.plus({months: 1});

  // Figure out the size of the left & right padding.
  const startOfmonth = month.startOf('month');
  const previousPad = weekdays.indexOf(startOfmonth.weekday);
  const rows = Math.ceil((month.daysInMonth + previousPad) / 7);
  const totalBoxes = rows * 7;
  const nextPad = totalBoxes - (previousPad + month.daysInMonth);

  // Build the day numbers for each range.
  const previousMonthDayNumbers = range(
    previousMonth.daysInMonth - previousPad + 1,
    previousMonth.daysInMonth + 1
  );
  const currentMonthDayNumbers = range(1, month.daysInMonth + 1);
  const nextMonthDayNumbers = range(1, nextPad + 1);

  // Map the day numbers to datetimes.
  const previousMonthDays = previousMonthDayNumbers.map((day: number) => previousMonth.set({day}));
  const currentMonthDays = currentMonthDayNumbers.map((day: number) => month.set({day}));
  const nextMonthDays = nextMonthDayNumbers.map((day: number) => nextMonth.set({day}));

  // Combine everything and return.
  return [...previousMonthDays, ...currentMonthDays, ...nextMonthDays];
}
