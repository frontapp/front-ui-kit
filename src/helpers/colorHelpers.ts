/*
 * Interfaces.
 */

interface FullColorPalette {
  readonly white: string;
  readonly shade10: string;
  readonly shade20: string;
  readonly shade30: string;
  readonly shade40: string;
  readonly shade50: string;
  readonly shade60: string;
  readonly shade70: string;
  readonly shade80: string;
  readonly shade90: string;
  readonly black: string;
}

export interface PaletteColor {
  readonly shade10: string;
  readonly shade20: string;
  readonly shade30: string;
  readonly shade40: string;
  readonly shade50: string;
}

type Palette = {readonly [T in PaletteColorsEnum]: PaletteColor};

/*
 * Constants.
 */

export enum PaletteColorsEnum {
  GREY = 'grey',
  PINK = 'pink',
  RED = 'red',
  ORANGE = 'orange',
  YELLOW = 'yellow',
  GREEN = 'green',
  TEAL = 'teal',
  BLUE = 'blue',
  PURPLE = 'purple'
}

export const greys: Readonly<FullColorPalette> = {
  white: '#ffffff',
  shade10: '#f7f8fa',
  shade20: '#f3f4f6',
  shade30: '#e8eaed',
  shade40: '#d4d6d9',
  shade50: '#bbbdbf',
  shade60: '#9b9c9e',
  shade70: '#77787a',
  shade80: '#3e3e40',
  shade90: '#19191a',
  black: '#001B38'
};

export const palette: Palette = {
  [PaletteColorsEnum.GREY]: {
    shade10: greys.shade10,
    shade20: greys.shade30,
    shade30: greys.shade50,
    shade40: greys.shade70,
    shade50: greys.shade80
  },
  [PaletteColorsEnum.PINK]: {
    shade10: '#fdf0f8',
    shade20: '#fbe1f2',
    shade30: '#f8cae8',
    shade40: '#ca3e99',
    shade50: '#9f3079'
  },
  [PaletteColorsEnum.RED]: {
    shade10: '#fff1f1',
    shade20: '#ffe1de',
    shade30: '#fac1bb',
    shade40: '#e9483a',
    shade50: '#b72215'
  },
  [PaletteColorsEnum.ORANGE]: {
    shade10: '#fff7f0',
    shade20: '#ffe9d1',
    shade30: '#fadab9',
    shade40: '#f2830b',
    shade50: '#cc6000'
  },
  [PaletteColorsEnum.YELLOW]: {
    shade10: '#fefbef',
    shade20: '#fef6dc',
    shade30: '#faebbb',
    shade40: '#e3b51e',
    shade50: '#91691b'
  },
  [PaletteColorsEnum.GREEN]: {
    shade10: '#eefbf3',
    shade20: '#d9f6e3',
    shade30: '#b6f0ca',
    shade40: '#10aa40',
    shade50: '#0c8331'
  },
  [PaletteColorsEnum.TEAL]: {
    shade10: '#f0fdff',
    shade20: '#dcfbff',
    shade30: '#b9f2fa',
    shade40: '#15acc0',
    shade50: '#0d7482'
  },
  [PaletteColorsEnum.BLUE]: {
    shade10: '#f1f6fd',
    shade20: '#e0edff',
    shade30: '#bed7fa',
    shade40: '#367fee',
    shade50: '#2356b3'
  },
  [PaletteColorsEnum.PURPLE]: {
    shade10: '#f8f1ff',
    shade20: '#f0dfff',
    shade30: '#ddbbfa',
    shade40: '#9235e4',
    shade50: '#681aad'
  }
};
