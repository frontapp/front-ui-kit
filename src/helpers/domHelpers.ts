export function getWidthWithMarginOfElement(node: Element) {
  if (!window || !window.document) return 0;

  const styles = window.getComputedStyle(node);
  const width = getPropertyAsNumber(styles, 'width') || 0;
  const marginLeft = getPropertyAsNumber(styles, 'margin-left') || 0;
  const marginRight = getPropertyAsNumber(styles, 'margin-right') || 0;

  return width + marginLeft + marginRight;
}

function getPropertyAsNumber(computedStyle: CSSStyleDeclaration, propertyName: string): number | null {
  const valueAsString: string = computedStyle.getPropertyValue(propertyName);
  const valueMatch = valueAsString.match(/(.*)px$/); // ComputedStyle calculates everything in pixels. e.g. '12.3px'.
  if (!valueMatch) return null;

  const value = Number(valueMatch[1]);
  return isNaN(value) ? null : value;
}
