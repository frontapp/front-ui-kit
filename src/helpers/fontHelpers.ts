/*
 * Interfaces.
 */

interface FontSizes {
  readonly veryTiny: string;
  readonly tiny: string;
  readonly verySmall: string;
  readonly small: string;
  readonly medium: string;
  readonly large: string;
  readonly veryLarge: string;
}

interface FontWeights {
  readonly thin: number;
  readonly light: number;
  readonly normal: number;
  readonly medium: number;
  readonly semibold: number;
  readonly bold: number;
  readonly black: number;
}

interface Fonts {
  readonly system: string;
  readonly monospace: string;
}

/*
 * Constants.
 */

export const fontSizes: FontSizes = {
  veryTiny: '10px',
  tiny: '11px',
  verySmall: '12px',
  small: '13px',
  medium: '14px',
  large: '16px',
  veryLarge: '18px'
};

export const fontWeights: FontWeights = {
  thin: 100,
  light: 300,
  normal: 400,
  medium: 500,
  semibold: 600,
  bold: 700,
  black: 800
};

export enum VisualSizesEnum {
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
  EXTRA_LARGE = 'extra-large'
}

export const fonts: Fonts = {
  system: `
    -apple-system,
    BlinkMacSystemFont,
    "Segoe UI",
    "Helvetica Neue",
    Helvetica,
    Arial,
    sans-serif,
    "Apple Color Emoji",
    "Segoe UI Emoji",
    "Segoe UI Symbol"
  `,
  monospace: `
    monaco,
    Consolas,
    "Lucida Console",
    monospace
  `
};
