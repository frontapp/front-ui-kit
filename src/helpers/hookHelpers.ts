import {useCallback, useEffect, useRef} from "react";

/*
 * Use Timeout.
 */

/** Returns a setTimeout-like function that clears the previous timeout when re-invoked or when the component unmounts. */
export function useTimeout(): [(handler: () => void, timeout?: number) => void, () => void] {
  const handleRef = useRef<number>();

  useEffect(() => () => clearTimeout(handleRef.current), []);

  const safeSetTimeout = useCallback((handler: () => void, timeout?: number) => {
    clearTimeout(handleRef.current);
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-explicit-any
    handleRef.current = setTimeout(handler, timeout) as any;
  }, []);
  const safeClearTimeout = useCallback(() => clearTimeout(handleRef.current), []);

  return [safeSetTimeout, safeClearTimeout];
}

/** Returns the passed value as it was after the last successful render of the component. */
export function usePrevious<T>(value: T) {
  const ref = useRef<T | null>(null);
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}
