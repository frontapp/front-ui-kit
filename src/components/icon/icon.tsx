import React, {FC} from 'react';
import styled from 'styled-components';

import CheckmarkCircle from '../../assets/icons/checkmarkCircle.svg';
import Checkmark from '../../assets/icons/checkmarkSelected.svg';
import ChevronDown from '../../assets/icons/chevronDown.svg';
import ChevronLeft from '../../assets/icons/chevronLeft.svg';
import ChevronRight from '../../assets/icons/chevronRight.svg';
import ChevronUp from '../../assets/icons/chevronUp.svg';
import Close from '../../assets/icons/close.svg';
import ExternalLink from '../../assets/icons/externalLink.svg';
import Minus from '../../assets/icons/minus.svg';
import EllipsisHorizontal from '../../assets/icons/moreHorizontal.svg';
import EllipsisVertical from '../../assets/icons/moreVertical.svg';
import Plus from '../../assets/icons/new.svg';
import PlusCircle from '../../assets/icons/plusCircle.svg';
import Search from '../../assets/icons/search.svg';
import {greys} from '../../helpers/colorHelpers';

/*
 * Constants.
 */

const defaultSize = 16;
const defaultColor = greys.shade50;

const importedIcons = {
  CheckmarkCircle,
  Checkmark,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  ChevronUp,
  Close,
  ExternalLink,
  Minus,
  EllipsisHorizontal,
  EllipsisVertical,
  Plus,
  PlusCircle,
  Search
};
export type IconName = keyof typeof importedIcons;
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
export const icons = importedIcons as {[name in IconName]: (props: React.SVGAttributes<SVGElement>) => React.ReactElement};

/*
 * Props.
 */

interface IconProps {
  /** The icon to display. */
  name: IconName;
  /** The color of the icon. */
  color?: string;
  /** Controls the width and height of the icon. */
  size?: number;
  /** If set, no color for the icon will be set. This enables inheriting the parents color. */
  shouldDisableColor?: boolean;
}

/*
 * Style.
 */

interface StyledIconDivProps {
  $size: number;
}

const StyledIconDiv = styled.div<StyledIconDivProps>`
  width: ${p => `${p.$size}px`};
  height: ${p => `${p.$size}px`};
`;

/*
 * Component.
 */

export const Icon: FC<IconProps> = props => {
  const {name, size = defaultSize, color = defaultColor, shouldDisableColor} = props;

  // Pull the icon from the list of available icons.
  // The svg file is converted to a function that is called.
  const icon = icons[name];
  if (!icon)
    return null;

  return (
    <StyledIconDiv $size={size}>
      {icon({
        width: size,
        height: size,
        color: shouldDisableColor ? '' : color,
        preserveAspectRatio: "none"
      })}
    </StyledIconDiv>
  );
};
