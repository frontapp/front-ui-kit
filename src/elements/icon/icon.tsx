import React, {FC} from 'react';
import styled from 'styled-components';

import Archive from '../../assets/icons/archive.svg';
import ArchiveFilled from '../../assets/icons/archiveFilled.svg';
import Assign from '../../assets/icons/assign.svg';
import AssignFilled from '../../assets/icons/assignFilled.svg';
import AttachmentArchive from '../../assets/icons/attachmentArchive.svg';
import AttachmentCalendar from '../../assets/icons/attachmentCalendar.svg';
import AttachmentCode from '../../assets/icons/attachmentCode.svg';
import AttachmentExcel from '../../assets/icons/attachmentExcel.svg';
import AttachmentGeneric from '../../assets/icons/attachmentGeneric.svg';
import AttachmentImage from '../../assets/icons/attachmentImage.svg';
import AttachmentMusic from '../../assets/icons/attachmentMusic.svg';
import AttachmentPdf from '../../assets/icons/attachmentPdf.svg';
import AttachmentPowerpoint from '../../assets/icons/attachmentPowerpoint.svg';
import AttachmentVideo from '../../assets/icons/attachmentVideo.svg';
import AttachmentWord from '../../assets/icons/attachmentWord.svg';
import Calendar from '../../assets/icons/calendar.svg';
import CalendarFilled from '../../assets/icons/calendarFilled.svg';
import CheckmarkBox from '../../assets/icons/checkmarkBox.svg';
import CheckmarkCircle from '../../assets/icons/checkmarkCircle.svg';
import CheckmarkCircleEmpty from '../../assets/icons/checkmarkCircleEmpty.svg';
import Checkmark from '../../assets/icons/checkmarkSelected.svg';
import ChevronDown from '../../assets/icons/chevronDown.svg';
import ChevronLeft from '../../assets/icons/chevronLeft.svg';
import ChevronRight from '../../assets/icons/chevronRight.svg';
import ChevronUp from '../../assets/icons/chevronUp.svg';
import Close from '../../assets/icons/close.svg';
import Copy from '../../assets/icons/copy.svg';
import Edit from '../../assets/icons/edit.svg';
import EditFilled from '../../assets/icons/editFilled.svg';
import Export from '../../assets/icons/export.svg';
import ExternalLink from '../../assets/icons/externalLink.svg';
import Gear from '../../assets/icons/gear.svg';
import GearFilled from '../../assets/icons/gearFilled.svg';
import Import from '../../assets/icons/import.svg';
import Info from '../../assets/icons/info.svg';
import InfoFilled from '../../assets/icons/infoFilled.svg';
import Minus from '../../assets/icons/minus.svg';
import EllipsisHorizontal from '../../assets/icons/moreHorizontal.svg';
import EllipsisVertical from '../../assets/icons/moreVertical.svg';
import Plus from '../../assets/icons/new.svg';
import NewContact from '../../assets/icons/newContact.svg';
import Participant from '../../assets/icons/participant.svg';
import ParticipantFilled from '../../assets/icons/participantFilled.svg';
import PlusCircle from '../../assets/icons/plusCircle.svg';
import Preferences from '../../assets/icons/preferences.svg';
import Search from '../../assets/icons/search.svg';
import Star from '../../assets/icons/star.svg';
import StarFilled from '../../assets/icons/starFilled.svg';
import Trash from '../../assets/icons/trash.svg';
import TrashFilled from '../../assets/icons/trashFilled.svg';
import WarningFilled from '../../assets/icons/warningFilled.svg';
import {greys} from '../../helpers/colorHelpers';

/*
 * Constants.
 */

const defaultSize = 16;
const defaultColor = greys.shade50;

const importedIcons = {
  Archive,
  ArchiveFilled,
  Assign,
  AssignFilled,
  AttachmentArchive,
  AttachmentCalendar,
  AttachmentCode,
  AttachmentExcel,
  AttachmentGeneric,
  AttachmentImage,
  AttachmentMusic,
  AttachmentPdf,
  AttachmentPowerpoint,
  AttachmentVideo,
  AttachmentWord,
  Calendar,
  CalendarFilled,
  CheckmarkCircle,
  CheckmarkCircleEmpty,
  CheckmarkBox,
  Checkmark,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  ChevronUp,
  Close,
  Copy,
  Edit,
  EditFilled,
  EllipsisHorizontal,
  EllipsisVertical,
  Export,
  ExternalLink,
  Gear,
  GearFilled,
  Import,
  Info,
  InfoFilled,
  Minus,
  NewContact,
  Participant,
  ParticipantFilled,
  Plus,
  PlusCircle,
  Preferences,
  Search,
  Star,
  StarFilled,
  Trash,
  TrashFilled,
  WarningFilled
};
export type IconName = keyof typeof importedIcons;
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
export const icons = importedIcons as {
  [name in IconName]: (props: React.SVGAttributes<SVGElement>) => React.ReactElement;
};

/*
 * Props.
 */

interface IconProps {
  /** The icon to display. */
  name: IconName;
  /** The color of the icon. */
  color?: string;
  /** Controls the width and height of the icon. */
  size?: number;
  /** If set, no color for the icon will be set. This enables inheriting the parents color. */
  shouldDisableColor?: boolean;
}

/*
 * Style.
 */

interface StyledIconDivProps {
  $size: number;
}

const StyledIconDiv = styled.div<StyledIconDivProps>`
  width: ${p => `${p.$size}px`};
  height: ${p => `${p.$size}px`};
`;

/*
 * Component.
 */

export const Icon: FC<IconProps> = props => {
  const {name, size = defaultSize, color = defaultColor, shouldDisableColor} = props;

  // Pull the icon from the list of available icons.
  // The svg file is converted to a function that is called.
  const icon = icons[name];
  if (!icon)
    return null;

  return (
    <StyledIconDiv $size={size}>
      {icon({
        name,
        width: size,
        height: size,
        color: shouldDisableColor ? '' : color,
        preserveAspectRatio: 'none'
      })}
    </StyledIconDiv>
  );
};
