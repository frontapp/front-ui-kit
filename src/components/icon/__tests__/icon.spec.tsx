import {render} from '@testing-library/react';
import React from 'react';

import {palette} from '../../../helpers/colorHelpers';
import {Icon} from '../../icon/icon';

describe('<Icon /> Spec', () => {
  describe('Snapshots', () => {
    it('should render with defaults', () => {
      const view = render(<Icon name="Checkmark" />);
      expect(view.container).toMatchSnapshot();
    });

    it('should support different sizes', () => {
      const view = render(<Icon name="Checkmark" size={32} />);
      expect(view.container).toMatchSnapshot();
    });

    it('should support different colors', () => {
      const view = render(<Icon name="Checkmark" color={palette.blue.shade40} />);
      expect(view.container).toMatchSnapshot();
    });
  });
});
