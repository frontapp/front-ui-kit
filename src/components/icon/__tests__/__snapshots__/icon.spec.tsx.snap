// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> Spec Snapshots should render with defaults 1`] = `
.c0 {
  width: 16px;
  height: 16px;
}

<div>
  <div
    class="c0"
  >
    <div
      color="#bbbdbf"
      height="16"
      name="Checkmark"
      preserveaspectratio="none"
      width="16"
    />
  </div>
</div>
`;

exports[`<Icon /> Spec Snapshots should support different colors 1`] = `
.c0 {
  width: 16px;
  height: 16px;
}

<div>
  <div
    class="c0"
  >
    <div
      color="#367fee"
      height="16"
      name="Checkmark"
      preserveaspectratio="none"
      width="16"
    />
  </div>
</div>
`;

exports[`<Icon /> Spec Snapshots should support different sizes 1`] = `
.c0 {
  width: 32px;
  height: 32px;
}

<div>
  <div
    class="c0"
  >
    <div
      color="#bbbdbf"
      height="32"
      name="Checkmark"
      preserveaspectratio="none"
      width="32"
    />
  </div>
</div>
`;
